jep/0000/README.adoc

```diff
@@ -0,0 +1,296 @@
+E.g. Bulk Loading the fingerprints for web UI.
+
+== Motivation
+
+File fingerprinting is a way to track which version of a file is being used by a job/build, making dependency tracking easy.
+The fingerprint engine of Jenkins can track usages of artifacts, credentials, files, etc. within the system.
+Currently, it does this by maintaining a local XML-based database which leads to dependence on the physical disk of the Jenkins master.
+
+Allowing fingerprint storage to be moved to external storages decreases the dependence of Jenkins instances on the physical disk space and also allows for tracking the flow of fingerprints across instances of Jenkins connected to the same external storage.
+
+External storages:
+
+* Open up the option of using pay-as-you-use cloud storages (often cheaper)
+* Make backup management easier
+* Ensure good availability and reliability
+
+== Reasoning
+
+=== Choice of Reference Implementation
+
+We choose an In-Memory DB: Redis due to the following reasons:
+
+* Since fingerprints are lightweight, they might be useful for users who don’t have a massive number of fingerprints and would benefit from the performance bump from IMDBs
+* Fast integration testing.
+* Popularity of Redis
+
+=== Direct agent to external storage interaction
+
+We decide not not to allow Jenkins agents to read/write fingerprint related information directly from/to the external storage without increasing load on the master.
+This is because the fingerprint data is lightweight and submission of the fingerprint back to the master involves just a small RPC packet.
+Therefore the load reduction may not be huge, plus there would be added code complexity as discussed link:https://docs.google.com/document/d/10f3IXTA6UMLUOFMTH_atQ3XlyWB3S7KGNCtTZmOUGdM/edit?disco=AAAAJMwkCMc[here]
+
+
+
+== Backwards Compatibility
+
+Backwards compatibility is highly important for the existing XML-based database to keep running smoothly.
+The proposal is designed accordingly, and explained in the Design section.
+
+== Security
+
+Fingerprints generally do contain sensitive information like artifacts, jobs, builds, etc. that we may want to hide from some set of users.
+
+We propose to add a word of caution to Jenkins admins that the external storage stores sensitive information and it is their responsibility to ensure security.
+
+[WARNING]
+====
+*(jglick)*
+I do not think cross-instance fingerprint storage (`loadAcrossInstancesById` etc.) can be considered safe. Access to job information is defined only by `AuthorizationStrategy` which applies only within one instance. Recommend deleting all cross-instance functionality.
+====
+
+
+This proposal does not define strong security requirements for external fingerprint storage implementations.
+
+== Infrastructure Requirements
+
+There are no new infrastructure requirements related to this proposal.
+
+== Testing
+
+Testing is relatively straightforward using JUnit, JTH, testcontainers, etc.
```

Comment:
Famous last words, but OK. It would be great to expand it later so that a reader could understand the scope